File: androidApp/src/main/kotlin/zinqq/app/screens/ChannelCloseDetailScreen.kt
package zinqq.app.screens

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import uniffi.wallet_core.CloseInitiatorView
import uniffi.wallet_core.CloseRecordView
import uniffi.wallet_core.CloseStatusLabel
import uniffi.wallet_core.CloseTxView
import zinqq.app.WalletHolder
import zinqq.app.blocksRemaining
import zinqq.app.closeAmountText
import zinqq.app.closeStatusLabel
import zinqq.app.closeTxRoleLabel
import zinqq.app.closeTypeLabel
import zinqq.app.components.CenteredNote
import zinqq.app.components.rememberCopiedFlash
import zinqq.app.confirmationText
import zinqq.app.explorerTxUrl
import zinqq.app.formatCloseDate
import zinqq.app.humanizeBlocks
import zinqq.app.isTerminalClose
import zinqq.app.midTruncate
import zinqq.app.nav.ScreenHeader
import zinqq.app.needsDeposit
import zinqq.app.theme.ZinqqTheme
import zinqq.app.totalFeesSats
import zinqq.main.formatBtc

/**
 * The PWA's ChannelCloseDetail (U14, R9/R11; `ChannelCloseDetail.tsx`):
 * live-updating dark room — status label, `~₿X` estimate while non-terminal,
 * the "Accessible in ~N days (N blocks)" countdown, a needs-deposit link to
 * Recover when the recovery state names this channel, fact rows, and the
 * per-tx list with role labels, confirmation counts, mempool links, and
 * 1,500ms copy feedback. Renders from `close_detail(channel_id)`, re-queried
 * on every wallet-data refresh.
 */
@Composable
fun ChannelCloseDetailScreen(
    holder: WalletHolder,
    channelId: String,
    onBack: () -> Unit,
    onRecover: () -> Unit,
) {
    val state by holder.state.collectAsState()
    val colors = ZinqqTheme.colors

    LaunchedEffect(channelId) { holder.loadCloseDetail(channelId) }
    val detail = state.closeDetail?.takeIf { it.channelId == channelId }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(colors.dark),
    ) {
        ScreenHeader(title = "Channel Close", onBack = onBack, tint = colors.onDark)

        when {
            detail == null -> CenteredNote("Loading...")
            detail.record == null -> CenteredNote("Close record not found")
            else -> CloseDetailBody(
                record = detail.record,
                needsDepositLink = needsDeposit(state.recoveryState, channelId),
                onRecover = onRecover,
            )
        }
    }
}

@Composable
private fun CloseDetailBody(
    record: CloseRecordView,
    needsDepositLink: Boolean,
    onRecover: () -> Unit,
) {
    val colors = ZinqqTheme.colors
    val terminal = isTerminalClose(record.status)
    val remaining = blocksRemaining(record)

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState()),
    ) {
        // Hero: live status label + estimated amount.
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 24.dp)
                .padding(top = 32.dp, bottom = 24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Text(
                text = closeStatusLabel(record.status),
                color = colors.onDarkMuted,
                fontSize = 18.sp,
                fontWeight = FontWeight.SemiBold,
            )
            Text(
                text = closeAmountText(record),
                color = colors.onDark,
                fontFamily = ZinqqTheme.fonts.display,
                fontWeight = FontWeight.Bold,
                fontSize = 36.sp,
                modifier = Modifier.padding(top = 8.dp),
            )
            if (!terminal) {
                Text(
                    text = buildString {
                        append(
                            if (record.initiator == CloseInitiatorView.REMOTE) {
                                "This channel was closed by the network. Your funds are safe " +
                                    "and return to your wallet automatically."
                            } else {
                                "Your funds return to your wallet automatically."
                            },
                        )
                        if (remaining != null && remaining > 0) {
                            append(
                                " Accessible in ${humanizeBlocks(remaining)} ($remaining blocks).",
                            )
                        }
                    },
                    color = colors.onDarkMuted,
                    fontSize = 14.sp,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(top = 8.dp),
                )
            }
            if (record.status == CloseStatusLabel.RESOLVED_UNVERIFIED) {
                Text(
                    text = "The close resolved on-chain, but this wallet couldn't verify " +
                        "receiving the funds — they may have been swept on another device.",
                    color = colors.warning,
                    fontSize = 14.sp,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(top = 8.dp),
                )
            }
        }

        if (needsDepositLink) {
            Text(
                text = "A small deposit is needed to recover these funds — tap to continue.",
                color = colors.warning,
                fontSize = 14.sp,
                modifier = Modifier
                    .padding(horizontal = 24.dp)
                    .padding(bottom = 16.dp)
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(8.dp))
                    .background(colors.warning.copy(alpha = 0.1f))
                    .clickable(onClick = onRecover)
                    .padding(12.dp),
            )
        }

        HorizontalDivider(
            color = colors.onDark.copy(alpha = 0.1f),
            modifier = Modifier.padding(horizontal = 24.dp),
        )

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 24.dp)
                .padding(top = 8.dp),
        ) {
            DetailRow(label = "Initiated", value = formatCloseDate(record.createdAtMs.toLong()))
            record.closureReason?.let { DetailRow(label = "Reason", value = it) }
            DetailRow(label = "Close type", value = closeTypeLabel(record.closeType))
            val totalFees = totalFeesSats(record)
            if (terminal && totalFees > 0) {
                DetailRow(label = "Total fees paid", value = formatBtc(totalFees))
            }
            record.completedAtMs?.let {
                DetailRow(label = "Completed", value = formatCloseDate(it.toLong()))
            }
        }

        if (record.txs.isNotEmpty()) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 24.dp)
                    .padding(top = 16.dp, bottom = 24.dp),
            ) {
                Text(
                    text = "Transactions",
                    color = colors.onDarkMuted,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Medium,
                )
                record.txs.forEachIndexed { index, tx ->
                    CloseTxRow(tx = tx, lastRow = index == record.txs.lastIndex)
                }
            }
        }
    }
}

@Composable
private fun CloseTxRow(tx: CloseTxView, lastRow: Boolean) {
    val colors = ZinqqTheme.colors
    val context = LocalContext.current
    val clipboard = LocalClipboardManager.current
    // The PWA's 1,500ms "Copied" flash (ChannelCloseDetail.tsx:56-59).
    var copied by rememberCopiedFlash(1_500)

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 12.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text = closeTxRoleLabel(tx.role),
                color = colors.onDark,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.weight(1f),
            )
            Text(
                text = confirmationText(tx),
                color = colors.onDarkMuted,
                fontSize = 12.sp,
            )
        }
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text = midTruncate(tx.txid, 10, 10, "…"),
                color = colors.onDark,
                fontFamily = FontFamily.Monospace,
                fontSize = 12.sp,
                textDecoration = TextDecoration.Underline,
                modifier = Modifier
                    .weight(1f)
                    .clickable {
                        context.startActivity(
                            Intent(Intent.ACTION_VIEW, Uri.parse(explorerTxUrl(tx.txid))),
                        )
                    },
            )
            Text(
                text = if (copied) "Copied" else "Copy txid",
                color = colors.onDark,
                fontSize = 12.sp,
                textDecoration = TextDecoration.Underline,
                modifier = Modifier
                    .padding(start = 8.dp)
                    .clickable {
                        clipboard.setText(AnnotatedString(tx.txid))
                        copied = true
                    },
            )
        }
        tx.feeSats?.let { fee ->
            Text(
                text = "Fee: ${formatBtc(fee.toLong())}",
                color = colors.onDarkMuted,
                fontSize = 12.sp,
                modifier = Modifier.padding(top = 4.dp),
            )
        }
        if (!lastRow) {
            HorizontalDivider(
                color = colors.onDark.copy(alpha = 0.1f),
                modifier = Modifier.padding(top = 12.dp),
            )
        }
    }
}

File: iosApp/iosApp/Screens/ChannelCloseDetailScreen.swift
import Shared
import SwiftUI
import UIKit

/// The PWA's ChannelCloseDetail (U19, R9/R11; `ChannelCloseDetail.tsx`),
/// mirroring Android's `ChannelCloseDetailScreen`: live-updating dark room —
/// status label, `~₿X` estimate while non-terminal, the "Accessible in ~N
/// days (N blocks)" countdown, a needs-deposit link to Recover when the
/// recovery state names this channel, fact rows, and the per-tx list with
/// role labels, confirmation counts, mempool links, and 1,500ms copy
/// feedback. Renders from `closeDetail(channelId)`, re-queried on every
/// wallet-data refresh.
struct ChannelCloseDetailScreen: View {
    @ObservedObject var model: WalletModel
    let channelId: String
    let onBack: () -> Void
    let onRecover: () -> Void

    @Environment(\.zinqqColors) private var colors

    var body: some View {
        let detail = model.closeDetail.flatMap { $0.channelId == channelId ? $0 : nil }

        VStack(spacing: 0) {
            ScreenHeader(title: "Channel Close", onBack: onBack, tint: colors.onDark)

            if let detail {
                if let record = detail.record {
                    CloseDetailBody(
                        record: record,
                        explorerBaseUrl: model.explorerBaseUrl,
                        needsDepositLink: needsDeposit(model.recoveryState, channelId: channelId),
                        onRecover: onRecover
                    )
                } else {
                    CenteredDarkNote("Close record not found")
                }
            } else {
                CenteredDarkNote("Loading...")
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(colors.dark.ignoresSafeArea())
        .task(id: channelId) {
            model.loadCloseDetail(channelId: channelId)
        }
    }
}

private struct CloseDetailBody: View {
    let record: CloseRecordView
    let explorerBaseUrl: String
    let needsDepositLink: Bool
    let onRecover: () -> Void

    @Environment(\.zinqqColors) private var colors

    var body: some View {
        let terminal = isTerminalClose(record.status)
        let remaining = blocksRemaining(record)

        ScrollView {
            VStack(spacing: 0) {
                // Hero: live status label + estimated amount.
                VStack(spacing: 8) {
                    Text(closeStatusLabel(record.status))
                        .font(ZinqqFont.sans(18, weight: .semibold))
                        .foregroundColor(colors.onDarkMuted)
                    Text(closeAmountText(record))
                        .font(ZinqqFont.display(36, weight: .bold))
                        .foregroundColor(colors.onDark)
                    if !terminal {
                        Text(nonTerminalCopy(remaining: remaining))
                            .font(ZinqqFont.sans(14))
                            .foregroundColor(colors.onDarkMuted)
                            .multilineTextAlignment(.center)
                    }
                    if record.status == .resolvedUnverified {
                        Text(
                            "The close resolved on-chain, but this wallet couldn't verify "
                                + "receiving the funds — they may have been swept on another device."
                        )
                        .font(ZinqqFont.sans(14))
                        .foregroundColor(colors.warning)
                        .multilineTextAlignment(.center)
                    }
                }
                .padding(.horizontal, 24)
                .padding(.top, 32)
                .padding(.bottom, 24)

                if needsDepositLink {
                    Button(action: onRecover) {
                        Text("A small deposit is needed to recover these funds — tap to continue.")
                            .font(ZinqqFont.sans(14))
                            .foregroundColor(colors.warning)
                            .multilineTextAlignment(.leading)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(12)
                            .background(colors.warning.opacity(0.1))
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                    .buttonStyle(.plain)
                    .padding(.horizontal, 24)
                    .padding(.bottom, 16)
                }

                Divider()
                    .overlay(colors.onDark.opacity(0.1))
                    .padding(.horizontal, 24)

                VStack(spacing: 0) {
                    DetailRow(
                        label: "Initiated",
                        value: formatCloseDate(timestampMs: Int64(bitPattern: record.createdAtMs))
                    )
                    if let reason = record.closureReason {
                        DetailRow(label: "Reason", value: reason)
                    }
                    DetailRow(label: "Close type", value: closeTypeLabel(record.closeType))
                    let totalFees = totalFeesSats(record)
                    if terminal && totalFees > 0 {
                        DetailRow(
                            label: "Total fees paid",
                            value: FormatKt.formatBtc(sats: totalFees)
                        )
                    }
                    if let completedAtMs = record.completedAtMs {
                        DetailRow(
                            label: "Completed",
                            value: formatCloseDate(timestampMs: completedAtMs.int64Value)
                        )
                    }
                }
                .padding(.horizontal, 24)
                .padding(.top, 8)

                if !record.txs.isEmpty {
                    VStack(alignment: .leading, spacing: 0) {
                        Text("Transactions")
                            .font(ZinqqFont.sans(14, weight: .medium))
                            .foregroundColor(colors.onDarkMuted)
                        ForEach(Array(record.txs.enumerated()), id: \.element.txid) { index, tx in
                            CloseTxRow(
                                explorerBaseUrl: explorerBaseUrl,
                                tx: tx,
                                lastRow: index == record.txs.count - 1
                            )
                        }
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, 24)
                    .padding(.top, 16)
                    .padding(.bottom, 24)
                }
            }
        }
    }

    private func nonTerminalCopy(remaining: Int64?) -> String {
        var copy = record.initiator == .remote
            ? "This channel was closed by the network. Your funds are safe "
                + "and return to your wallet automatically."
            : "Your funds return to your wallet automatically."
        if let remaining, remaining > 0 {
            copy += " Accessible in \(humanizeBlocks(remaining)) (\(remaining) blocks)."
        }
        return copy
    }
}

private struct CloseTxRow: View {
    let explorerBaseUrl: String
    let tx: CloseTxView
    let lastRow: Bool

    @Environment(\.zinqqColors) private var colors
    @State private var copied = false

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack {
                Text(closeTxRoleLabel(tx.role))
                    .font(ZinqqFont.sans(14, weight: .semibold))
                    .foregroundColor(colors.onDark)
                    .frame(maxWidth: .infinity, alignment: .leading)
                Text(confirmationText(tx))
                    .font(ZinqqFont.sans(12))
                    .foregroundColor(colors.onDarkMuted)
            }
            HStack {
                Button {
                    if let url = URL(string: explorerTxUrl(explorerBaseUrl, tx.txid)) {
                        UIApplication.shared.open(url)
                    }
                } label: {
                    Text(midTruncate(tx.txid, head: 10, tail: 10, ellipsis: "…"))
                        .font(.system(size: 12, design: .monospaced))
                        .foregroundColor(colors.onDark)
                        .underline()
                }
                .accessibilityLabel("View transaction on mempool.space")
                .frame(maxWidth: .infinity, alignment: .leading)

                Button {
                    UIPasteboard.general.string = tx.txid
                    copied = true
                } label: {
                    Text(copied ? "Copied" : "Copy txid")
                        .font(ZinqqFont.sans(12))
                        .foregroundColor(colors.onDark)
                        .underline()
                }
                .accessibilityLabel("Copy transaction id")
                .padding(.leading, 8)
            }
            // The PWA's 1,500ms "Copied" flash (ChannelCloseDetail.tsx:56-59).
            .autoReset($copied, afterMs: 1_500)
            if let fee = tx.feeSats?.int64Value {
                Text("Fee: \(FormatKt.formatBtc(sats: fee))")
                    .font(ZinqqFont.sans(12))
                    .foregroundColor(colors.onDarkMuted)
            }
            if !lastRow {
                Divider()
                    .overlay(colors.onDark.opacity(0.1))
                    .padding(.top, 12)
            }
        }
        .padding(.vertical, 12)
    }
}
